File: test/TestBasicEff.java
package com.tann.dice.test;

import com.tann.dice.gameplay.content.ent.Hero;
import com.tann.dice.gameplay.content.ent.die.side.blob.ESB;
import com.tann.dice.gameplay.content.ent.type.HeroType;
import com.tann.dice.gameplay.content.ent.type.MonsterType;
import com.tann.dice.gameplay.content.ent.type.lib.HeroTypeUtils;
import com.tann.dice.gameplay.content.ent.type.lib.MonsterTypeLib;
import com.tann.dice.gameplay.effect.eff.EffBill;
import com.tann.dice.gameplay.fightLog.EntState;
import com.tann.dice.gameplay.fightLog.FightLog;
import com.tann.dice.test.util.Test;
import com.tann.dice.test.util.TestRunner;
import com.tann.dice.test.util.TestUtils;

public class TestBasicEff {
   @Test
   public static void basicSanityTest() {
      FightLog f = TestUtils.setupFight();
      TestRunner.assertEquals("There should be 1 monster", 1, f.getSnapshot(FightLog.Temporality.Present).getAliveMonsterStates().size());
      TestRunner.assertEquals("There should be 1 hero", 1, f.getSnapshot(FightLog.Temporality.Present).getAliveHeroStates().size());
   }

   @Test
   public static void attackEnemy() {
      FightLog f = TestUtils.setupFight();
      EntState monsterState = f.getState(FightLog.Temporality.Future, TestUtils.monsters.get(0));
      TestRunner.assertEquals("Monster should be undamaged", monsterState.getMaxHp(), monsterState.getHp());
      TestUtils.attack(f, TestUtils.heroes.get(0), TestUtils.monsters.get(0), 1, false);
      monsterState = f.getState(FightLog.Temporality.Future, TestUtils.monsters.get(0));
      TestRunner.assertEquals("Monster should damaged", monsterState.getMaxHp() - 1, monsterState.getHp());
   }

   @Test
   public static void attackHero() {
      FightLog f = TestUtils.setupFight();
      EntState heroState = f.getState(FightLog.Temporality.Future, TestUtils.heroes.get(0));
      TestRunner.assertEquals("Hero should be undamaged", heroState.getMaxHp(), heroState.getHp());
      TestUtils.attack(f, TestUtils.monsters.get(0), TestUtils.heroes.get(0), 1, false);
      heroState = f.getState(FightLog.Temporality.Future, TestUtils.heroes.get(0));
      TestRunner.assertEquals("Hero should damaged", heroState.getMaxHp() - 1, heroState.getHp());
   }

   @Test
   public static void basicBlock() {
      FightLog f = TestUtils.setupFight();
      Hero hero = TestUtils.heroes.get(0);
      TestUtils.hit(f, hero, hero, ESB.shield.val(1), false);
      TestUtils.hit(f, hero, hero, ESB.dmg.val(2), false);
      EntState heroState = f.getState(FightLog.Temporality.Future, TestUtils.heroes.get(0));
      TestRunner.assertEquals("Hero should take 1 damage", heroState.getMaxHp() - 1, heroState.getHp());
      TestRunner.assertEquals("Hero should block 1 damage", 1, heroState.getDamageBlocked());
   }

   @Test
   public static void basicHeal() {
      FightLog f = TestUtils.setupFight();
      Hero hero = TestUtils.heroes.get(0);
      TestUtils.hit(f, hero, hero, ESB.dmg.val(2), false);
      TestUtils.hit(f, hero, hero, ESB.heal.val(3), false);
      EntState heroState = f.getState(FightLog.Temporality.Future, TestUtils.heroes.get(0));
      TestRunner.assertEquals("Hero should be on full hp", heroState.getMaxHp(), heroState.getHp());
   }

   @Test
   public static void reinforcements() {
      MonsterType[] mt = new MonsterType[30];

      for (int i = 0; i < mt.length; i++) {
         mt[i] = MonsterTypeLib.byName("testGoblin");
      }

      FightLog f = TestUtils.setupFight(new HeroType[]{HeroTypeUtils.byName("Healer")}, mt);
      int initialGoblinsPresent = f.getSnapshot(FightLog.Temporality.Present).getEntities(false, false).size();
      TestRunner.assertEquals("should be a few goblins at the start", true, initialGoblinsPresent > 3 && initialGoblinsPresent < 20);
      TestUtils.hit(f, null, new EffBill().damage(100).bEff(), false);
      TestRunner.assertEquals(
         "goblins present should not have changed", initialGoblinsPresent, f.getSnapshot(FightLog.Temporality.Present).getEntities(false, false).size()
      );
      TestUtils.hit(f, null, new EffBill().damage(100).group().bEff(), false);
      TestRunner.assertEquals(
         "goblins present should not have changed", initialGoblinsPresent, f.getSnapshot(FightLog.Temporality.Present).getEntities(false, false).size()
      );

      for (int i = 0; i < 8; i++) {
         TestUtils.hit(f, null, new EffBill().damage(100).group().bEff(), false);
      }

      TestRunner.assertEquals("goblins present should be 0", 0, f.getSnapshot(FightLog.Temporality.Present).getEntities(false, false).size());
      TestRunner.assertEquals("should be victorious", true, f.getSnapshot(FightLog.Temporality.Present).isVictory());
   }
}

File: combat/tests/test_basic_eff.py
"""Tests from TestBasicEff.java - basic effect tests."""

from src.entity import Entity, Team, FIGHTER, GOBLIN, HEALER, TEST_GOBLIN
from src.fight import FightLog, Temporality


def setup_fight() -> tuple[FightLog, Entity, Entity]:
    """Standard 1v1 fight setup: 1 Fighter vs 1 Goblin.

    Returns (fight, hero, monster) tuple for convenience.
    Matches Java's TestUtils.setupFight() default behavior.
    """
    heroes = [Entity(FIGHTER, Team.HERO, 0)]
    monsters = [Entity(GOBLIN, Team.MONSTER, 0)]
    fight = FightLog(heroes, monsters)
    return fight, heroes[0], monsters[0]


def test_basic_sanity():
    """setupFight creates exactly 1 hero and 1 monster.

    From TestBasicEff.basicSanityTest - verifies test harness works.
    """
    fight, hero, monster = setup_fight()

    # Check alive counts in present state
    alive_heroes = fight.get_alive_heroes(Temporality.PRESENT)
    alive_monsters = fight.get_alive_monsters(Temporality.PRESENT)

    assert len(alive_heroes) == 1, "There should be 1 hero"
    assert len(alive_monsters) == 1, "There should be 1 monster"


def test_attack_enemy():
    """Attacking a monster reduces its HP by the damage amount.

    From TestBasicEff.attackEnemy - basic damage to enemy.
    """
    fight, hero, monster = setup_fight()

    # Monster should start undamaged
    monster_state = fight.get_state(monster, Temporality.FUTURE)
    assert monster_state.hp == monster_state.max_hp, "Monster should be undamaged"

    # Hero attacks monster for 1 damage (immediate, not pending)
    fight.apply_damage(hero, monster, amount=1, is_pending=False)

    # Monster should now have 1 less HP
    monster_state = fight.get_state(monster, Temporality.FUTURE)
    assert monster_state.hp == monster_state.max_hp - 1, "Monster should be damaged"


def test_attack_hero():
    """Attacking a hero reduces their HP by the damage amount.

    From TestBasicEff.attackHero - basic damage to hero.
    """
    fight, hero, monster = setup_fight()

    # Hero should start undamaged
    hero_state = fight.get_state(hero, Temporality.FUTURE)
    assert hero_state.hp == hero_state.max_hp, "Hero should be undamaged"

    # Monster attacks hero for 1 damage (immediate, not pending)
    fight.apply_damage(monster, hero, amount=1, is_pending=False)

    # Hero should now have 1 less HP
    hero_state = fight.get_state(hero, Temporality.FUTURE)
    assert hero_state.hp == hero_state.max_hp - 1, "Hero should be damaged"


def test_basic_block():
    """Shield blocks incoming damage.

    From TestBasicEff.basicBlock - shield(1) + dmg(2) = 1 damage taken, 1 blocked.
    """
    fight, hero, monster = setup_fight()

    # Give hero 1 shield
    fight.apply_shield(hero, amount=1)

    # Take 2 damage (immediate)
    fight.apply_damage(monster, hero, amount=2, is_pending=False)

    # Hero should have taken only 1 damage (2 - 1 blocked)
    hero_state = fight.get_state(hero, Temporality.FUTURE)
    assert hero_state.hp == hero_state.max_hp - 1, "Hero should take 1 damage"
    assert hero_state.damage_blocked == 1, "Hero should have blocked 1 damage"


def test_basic_heal():
    """Heal restores HP, capped at max HP.

    From TestBasicEff.basicHeal - dmg(2) + heal(3) = full HP (capped).
    """
    fight, hero, monster = setup_fight()

    # Take 2 damage
    fight.apply_damage(monster, hero, amount=2, is_pending=False)

    # Verify damaged
    hero_state = fight.get_state(hero, Temporality.FUTURE)
    assert hero_state.hp == hero_state.max_hp - 2, "Hero should be damaged"

    # Heal for 3 (more than the 2 damage)
    fight.apply_heal(hero, amount=3)

    # Hero should be back to full HP (heal capped at max)
    hero_state = fight.get_state(hero, Temporality.FUTURE)
    assert hero_state.hp == hero_state.max_hp, "Hero should be on full hp"


def test_reinforcements():
    """Large monster groups spawn in waves based on field capacity.

    From TestBasicEff.reinforcements - tests the reinforcement spawning system.

    Field capacity: 165 units
    Hero-sized (testGoblin): 24 units each
    Max on field: floor(165/24) = 6

    Reinforcements spawn immediately when room becomes available (monster dies).
    """
    # Create 30 testGoblins
    monsters = [Entity(TEST_GOBLIN, Team.MONSTER, i) for i in range(30)]
    heroes = [Entity(HEALER, Team.HERO, 0)]
    fight = FightLog(heroes, monsters)

    # Should have 3-20 goblins present initially (spec says 6 fit)
    initial_present = len(fight.get_present_monsters(Temporality.PRESENT))
    assert initial_present > 3 and initial_present < 20, \
        f"Should be a few goblins at the start, got {initial_present}"

    # More specifically: 6 should fit (6*24=144 <= 165, 7*24=168 > 165)
    assert initial_present == 6, f"Expected 6 goblins to fit, got {initial_present}"

    # Single target damage (kills one, reinforcement spawns)
    target = fight.get_present_monsters(Temporality.PRESENT)[0]
    fight.apply_damage(None, target, amount=100, is_pending=False)

    # Count should not have changed (killed 1, spawned 1)
    present_after_single = len(fight.get_present_monsters(Temporality.PRESENT))
    assert present_after_single == initial_present, \
        f"Goblins present should not have changed after single damage, got {present_after_single}"

    # Group damage (kills all present, spawns replacements)
    fight.apply_group_damage(None, amount=100, is_pending=False)

    # Count should not have changed (as long as reinforcements remain)
    present_after_group = len(fight.get_present_monsters(Temporality.PRESENT))
    assert present_after_group == initial_present, \
        f"Goblins present should not have changed after group damage, got {present_after_group}"

    # Keep dealing group damage until all 30 are dead
    # We've killed: 1 (single) + 6 (group) = 7
    # Remaining: 30 - 7 = 23
    # Each group kills 6, so need ceil(23/6) = 4 more groups
    for _ in range(8):  # Match Java test (does more than needed)
        fight.apply_group_damage(None, amount=100, is_pending=False)

    # All goblins should be dead now
    present_final = len(fight.get_present_monsters(Temporality.PRESENT))
    assert present_final == 0, f"Goblins present should be 0, got {present_final}"

    # Should be victorious
    assert fight.is_victory(Temporality.PRESENT), "Should be victorious"
